package-res/resources/web/pentaho/type/complex.js

```diff
@@ -608,6 +608,46 @@ define([
       //endregion
       //endregion
 
+      //region serialization
+      serializeComplexValues: function(spec, values, keyArgs){
+        for (var property in values) {
+          if (values.hasOwnProperty(property)) {
+            if(values[property]._values){
```

Comment:
Each value should know how to serialize itself...
It would be even better to pass the control of serialization to an auxiliary method of each property, like Property.Meta#toSpecInnerInstance. This is similar to what was done for Property.Meta.validate, https://github.com/pentaho/pentaho-platform-plugin-common-ui/blob/master/package-res/resources/web/pentaho/type/Property.js#L441.

So, in the end, List#toSpecInner, Simple#toSpecInner and Complex#toSpecInner will each have its own code and you would not put _here_ the code for each case.